planner/implementation/datasource.go

```diff
@@ -171,3 +171,87 @@ func NewIndexScanImpl(scan *plannercore.PhysicalIndexScan, tblColHists *statisti
 		tblColHists: tblColHists,
 	}
 }
+
+// IndexLookUpReaderImpl is the Implementation of PhysicalIndexLookUpReader.
+type IndexLookUpReaderImpl struct {
+	baseImpl
+
+	KeepOrder          bool
+	DoubleReadNeedProj bool
+	tblColHists        *statistics.HistColl
+}
+
+// NewIndexLookUpReaderImpl creates a new table reader Implementation.
+func NewIndexLookUpReaderImpl(reader plannercore.PhysicalPlan, hists *statistics.HistColl) *IndexLookUpReaderImpl {
+	base := baseImpl{plan: reader}
+	impl := &IndexLookUpReaderImpl{
+		baseImpl:    base,
+		tblColHists: hists,
+	}
+	return impl
+}
+
+// CalcCost calculates the cost of the table reader Implementation.
+func (impl *IndexLookUpReaderImpl) CalcCost(outCount float64, children ...memo.Implementation) float64 {
```

Comment:
Can we reuse the code?

Reply:
Not easy for that, if we force to reuse, there are only little codes that can be reused.